Terms in _id Field are not correct UTF-8

<!--

** Please read the guidelines below. **

Issues that do not follow these guidelines are likely to be closed.

1.  GitHub is reserved for bug reports and feature requests. The best place to
    ask a general question is at the Elastic [forums](https://discuss.elastic.co).
    GitHub is not the place for general questions.

2.  Is this bug report or feature request for a supported OS? If not, it
    is likely to be closed.  See https://www.elastic.co/support/matrix#show_os

3.  Please fill out EITHER the feature request block or the bug report block
    below, and delete the other block.

-->

<!-- Bug report -->

**Elasticsearch version** (`bin/elasticsearch --version`): 6.5.0

**Plugins installed**: none

**JVM version** (`java -version`): 1.8.0_126

**OS version** (`uname -a` if on a Unix-like system): Windows 10

**Description of the problem including expected versus actual behavior**:

It seems that terms in the _id field are not correct UTF-8. This causes problems when Lucene tries to generate a char[] or a String for these terms eg, with UnicodeUtil.UTF8toUTF16.

**Steps to reproduce**:

Please include a *minimal* but *complete* recreation of the problem, including
(e.g.) index creation, mappings, settings, query etc.  The easier you make for
us to reproduce it, the more likely that somebody will take the time to look at it.

 1. Create Index

PUT my_index
{
  "settings": {
    "number_of_shards" : 1
  },
    "mappings" : {
    "_doc": {
      "properties" : {
        "body" : {
          "type" : "text", 
          "index" : true, 
          "analyzer": "standard",  
          "search_analyzer": "standard",
          "index_options": "offsets"
        }
      }
    }
  }
}

 2. Insert Document

PUT /my_index/_doc/1
{
  "body":  "aaa bbb ccc ddd eee fff ggg"
}

 3. Use UnifiedHighlighter with a query that enforces Analysis-Mode and a filter for the _id fields

GET /my_index/_search
{
    "query" : {
        "bool": {
          "must": [ 
            {
            "span_containing" : {
              "little" : {
                "span_term" : { "body" : "ccc" }
              },
              "big" : {
                "span_near" : {
                    "clauses" : [
                        { "span_term" : { "body" : "bbb" } },
                        { "span_term" : { "body" : "ddd" } }
                    ],
                    "slop" : 5,
                    "in_order" : true
                }
              }
            }
            }
          ],
			    "filter": [
            	{ "term": { "_id": "1" } }
          ]	
        }
      },
	 "_source": {
	    "includes": ["body"],
	    "excludes": []
	  },
     "highlight" : {
      "fields": {
		  "body": {
			"type": "unified",
			"number_of_fragments" : 10,
			"fragment_size" : 100
			}
		},
    "require_field_match": false
	}
}

**Provide logs (if relevant)**:

Caused by: java.lang.ArrayIndexOutOfBoundsException: 2
	at org.apache.lucene.util.UnicodeUtil.UTF8toUTF16(UnicodeUtil.java:602) ~[lucene-core-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:01:13]
	at org.apache.lucene.util.UnicodeUtil.UTF8toUTF16(UnicodeUtil.java:626) ~[lucene-core-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:01:13]
	at org.apache.lucene.util.automaton.DaciukMihovAutomatonBuilder.build(DaciukMihovAutomatonBuilder.java:303) ~[lucene-core-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:01:13]
	at org.apache.lucene.util.automaton.Automata.makeStringUnion(Automata.java:581) ~[lucene-core-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:01:13]
	at org.apache.lucene.search.uhighlight.MemoryIndexOffsetStrategy.buildCombinedAutomaton(MemoryIndexOffsetStrategy.java:67) ~[lucene-highlighter-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.apache.lucene.search.uhighlight.MemoryIndexOffsetStrategy.<init>(MemoryIndexOffsetStrategy.java:57) ~[lucene-highlighter-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.apache.lucene.search.uhighlight.UnifiedHighlighter.getOffsetStrategy(UnifiedHighlighter.java:866) ~[lucene-highlighter-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.apache.lucene.search.uhighlight.CustomUnifiedHighlighter.getFieldHighlighter(CustomUnifiedHighlighter.java:149) ~[elasticsearch-6.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.apache.lucene.search.uhighlight.UnifiedHighlighter.highlightFieldsAsObjects(UnifiedHighlighter.java:574) ~[lucene-highlighter-7.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.apache.lucene.search.uhighlight.CustomUnifiedHighlighter.highlightField(CustomUnifiedHighlighter.java:107) ~[elasticsearch-6.5.0.jar:7.5.0 b5bf70b7e32d7ddd9742cc821d471c5fabd4e3df - jimczi - 2018-09-18 13:02:19]
	at org.elasticsearch.search.fetch.subphase.highlight.UnifiedHighlighter.highlight(UnifiedHighlighter.java:128) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.fetch.subphase.highlight.HighlightPhase.hitExecute(HighlightPhase.java:110) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.fetch.FetchPhase.execute(FetchPhase.java:159) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.SearchService.executeFetchPhase(SearchService.java:413) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.SearchService.executeQueryPhase(SearchService.java:388) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.SearchService.access$100(SearchService.java:121) ~[elasticsearch-6.5.0.jar:6.5.0]
	at org.elasticsearch.search.SearchService$2.onResponse(SearchService.java:339) ~[elasticsearch-6.5.0.jar:6.5.0]
	... 9 more

--- a/server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java
+++ b/server/src/main/java/org/elasticsearch/search/fetch/subphase/highlight/UnifiedHighlighter.java
@@ -35,6 +35,7 @@ import org.elasticsearch.common.Strings;
 import org.elasticsearch.common.text.Text;
 import org.elasticsearch.index.IndexSettings;
 import org.elasticsearch.index.mapper.DocumentMapper;
+import org.elasticsearch.index.mapper.IdFieldMapper;
 import org.elasticsearch.index.mapper.MappedFieldType;
 import org.elasticsearch.search.fetch.FetchPhaseExecutionException;
 import org.elasticsearch.search.fetch.FetchSubPhase;
@@ -109,7 +110,9 @@ public class UnifiedHighlighter implements Highlighter {
                 final String fieldName = highlighterContext.fieldName;
                 highlighter.setFieldMatcher((name) -> fieldName.equals(name));
             } else {
-                highlighter.setFieldMatcher((name) -> true);
+                // ignore terms that targets the _id field since they use a different encoding
+                // that is not compatible with utf8
+                highlighter.setFieldMatcher(name -> IdFieldMapper.NAME.equals(name) == false);
             }
 
             Snippet[] fieldSnippets = highlighter.highlightField(highlighterContext.fieldName,

--- a/server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java
+++ b/server/src/test/java/org/elasticsearch/search/fetch/subphase/highlight/HighlighterSearchIT.java
@@ -2947,6 +2947,29 @@ public class HighlighterSearchIT extends ESIntegTestCase {
         }
     }
 
+    public void testDisableHighlightIdField() throws Exception {
+        assertAcked(prepareCreate("test")
+            .addMapping("doc", "keyword", "type=keyword"));
+        ensureGreen();
+
+        client().prepareIndex("test", "doc", "d33f85bf1e51e84d9ab38948db9f3a068e1fe5294f1d8603914ac8c7bcc39ca1")
+            .setSource("keyword", "Hello World")
+            .setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE)
+            .get();
+
+        for (String highlighterType : new String[] {"plain", "unified"}) {
+            SearchResponse searchResponse = client().prepareSearch()
+                .setQuery(matchQuery("_id", "d33f85bf1e51e84d9ab38948db9f3a068e1fe5294f1d8603914ac8c7bcc39ca1"))
+                .highlighter(
+                    new HighlightBuilder().field(new Field("*")
+                        .highlighterType(highlighterType).requireFieldMatch(false))
+                )
+                .get();
+            assertHitCount(searchResponse, 1);
+            assertNull(searchResponse.getHits().getAt(0).getHighlightFields().get("_id"));
+        }
+    }
+
     public static class MockAnalysisPlugin extends Plugin implements AnalysisPlugin {
 
         public final class MockSnowBall extends TokenFilter {

